pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/ColumnValueSegmentPruner.java

```diff
@@ -0,0 +1,211 @@
+import com.linkedin.pinot.common.utils.request.RequestUtils;
+import com.linkedin.pinot.core.common.predicate.RangePredicate;
+import com.linkedin.pinot.core.indexsegment.IndexSegment;
+import com.linkedin.pinot.core.segment.index.ColumnMetadata;
+import com.linkedin.pinot.core.segment.index.SegmentMetadataImpl;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import org.apache.commons.configuration.Configuration;
+
+
+/**
+ * An implementation of SegmentPruner.
+ * <p>Pruner will prune segment based on the column value inside the filter.
+ */
+public class ColumnValueSegmentPruner implements SegmentPruner {
+
+  @Override
+  public void init(Configuration config) {
+  }
+
+  @Override
+  public boolean prune(@Nonnull IndexSegment segment, @Nonnull BrokerRequest brokerRequest) {
+    FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
+    if (filterQueryTree == null) {
+      return false;
+    }
+    // For realtime segment, this map can be null.
+    Map<String, ColumnMetadata> columnMetadataMap =
+        ((SegmentMetadataImpl) segment.getSegmentMetadata()).getColumnMetadataMap();
+    return (columnMetadataMap != null) && pruneSegment(filterQueryTree, columnMetadataMap);
+  }
+
+  @Override
+  public String toString() {
+    return "ColumnValueSegmentPruner";
+  }
+
+  /**
+   * Helper method to determine if a segment can be pruned based on the column min/max value in segment metadata and
+   * the predicates on time column. The algorithm is as follows:
+   *
+   * <ul>
+   *   <li> For leaf node: Returns true if there is a predicate on the column and apply the predicate would result in
+   *   filtering out all docs of the segment, false otherwise. </li>
+   *   <li> For non-leaf AND node: True if any of its children returned true, false otherwise. </li>
+   *   <li> For non-leaf OR node: True if all its children returned true, false otherwise. </li>
+   * </ul>
+   *
+   * @param filterQueryTree Filter tree for the query.
+   * @param columnMetadataMap Map from column name to column metadata.
+   * @return True if segment can be pruned out, false otherwise.
+   */
+  @SuppressWarnings("unchecked")
+  public static boolean pruneSegment(@Nonnull FilterQueryTree filterQueryTree,
+      @Nonnull Map<String, ColumnMetadata> columnMetadataMap) {
+    FilterOperator filterOperator = filterQueryTree.getOperator();
+    List<FilterQueryTree> children = filterQueryTree.getChildren();
+
+    if (children == null || children.isEmpty()) {
+      // Leaf Node
+
+      // Skip operator other than EQUALITY and RANGE
+      if ((filterOperator != FilterOperator.EQUALITY) && (filterOperator != FilterOperator.RANGE)) {
+        return false;
+      }
+
+      ColumnMetadata columnMetadata = columnMetadataMap.get(filterQueryTree.getColumn());
+      if (columnMetadata == null) {
+        // Should not reach here after DataSchemaSegmentPruner
```

Comment:
There is no guarantee on the order of pruner right now, so we just prune the segment when reaching here.